epochStart/bootstrap/shardStorageHandler.go

```diff
@@ -604,75 +604,76 @@ func getEpochShardDataAndNeededMetaBlock(
 	return epochShardData, neededMeta, nil
 }
 
-func getMiniBlocksInfo(epochShardData data.EpochStartShardDataHandler, neededMeta *block.MetaBlock, shardID uint32) *miniBlockInfo {
-	mbInfo := &miniBlockInfo{
+func getMiniBlocksInfo(epochShardData data.EpochStartShardDataHandler, neededMeta *block.MetaBlock, shardID uint32) *miniBlocksInfo {
+	mbsInfo := &miniBlocksInfo{
 		miniBlockHashes:              make([][]byte, 0),
 		fullyProcessed:               make([]bool, 0),
 		indexOfLastTxProcessed:       make([]int32, 0),
 		pendingMiniBlocksMap:         make(map[string]struct{}),
 		pendingMiniBlocksPerShardMap: make(map[uint32][][]byte),
 	}
 
-	setMiniBlockInfoWithPendingMiniBlocks(epochShardData, mbInfo)
-	setMiniBlockInfoWithProcessedMiniBlocks(neededMeta, shardID, mbInfo)
+	setMiniBlocksInfoWithPendingMiniBlocks(epochShardData, mbsInfo)
+	setMiniBlocksInfoWithProcessedMiniBlocks(neededMeta, shardID, mbsInfo)
 
-	return mbInfo
+	return mbsInfo
 }
 
-func setMiniBlockInfoWithPendingMiniBlocks(epochShardData data.EpochStartShardDataHandler, mbInfo *miniBlockInfo) {
+func setMiniBlocksInfoWithPendingMiniBlocks(epochShardData data.EpochStartShardDataHandler, mbsInfo *miniBlocksInfo) {
 	for _, mbHeader := range epochShardData.GetPendingMiniBlockHeaderHandlers() {
-		log.Debug("shardStorageHandler.setMiniBlockInfoWithPendingMiniBlocks",
+		log.Debug("shardStorageHandler.setMiniBlocksInfoWithPendingMiniBlocks",
 			"mb hash", mbHeader.GetHash(),
 			"len(reserved)", len(mbHeader.GetReserved()),
 			"index of first tx processed", mbHeader.GetIndexOfFirstTxProcessed(),
 			"index of last tx processed", mbHeader.GetIndexOfLastTxProcessed(),
 		)
 
 		receiverShardID := mbHeader.GetReceiverShardID()
-		mbInfo.pendingMiniBlocksPerShardMap[receiverShardID] = append(mbInfo.pendingMiniBlocksPerShardMap[receiverShardID], mbHeader.GetHash())
-		mbInfo.pendingMiniBlocksMap[string(mbHeader.GetHash())] = struct{}{}
-
-		if mbHeader.GetIndexOfLastTxProcessed() > -1 {
-			mbInfo.miniBlockHashes = append(mbInfo.miniBlockHashes, mbHeader.GetHash())
-			mbInfo.fullyProcessed = append(mbInfo.fullyProcessed, false)
-			mbInfo.indexOfLastTxProcessed = append(mbInfo.indexOfLastTxProcessed, mbHeader.GetIndexOfLastTxProcessed())
+		mbsInfo.pendingMiniBlocksPerShardMap[receiverShardID] = append(mbsInfo.pendingMiniBlocksPerShardMap[receiverShardID], mbHeader.GetHash())
+		mbsInfo.pendingMiniBlocksMap[string(mbHeader.GetHash())] = struct{}{}
+
+		isPendingMiniBlockPartiallyExecuted := mbHeader.GetIndexOfLastTxProcessed() > -1 && mbHeader.GetIndexOfLastTxProcessed() < int32(mbHeader.GetTxCount())-1
```

Comment:
so this composed condition will prevent the current edge case. Can we still have the scenario in which both fields are 0 (due to the unmarshaling of the empty slice) in which the if branch will still be executed?

Reply:
No, because if reserved will be nil, no matter of reason, the first line in the method returns TxCount-1

```
// GetIndexOfLastTxProcessed returns index of the last transaction processed in the miniBlock
func (m *MiniBlockHeader) GetIndexOfLastTxProcessed() int32 {
	miniBlockHeaderReserved, err := m.getMiniBlockHeaderReserved()
	if err != nil || miniBlockHeaderReserved == nil {
		return int32(m.TxCount) - 1
	}
```

Reply:
And here we check only with GetIndexOfLastTxProcessed twice and not with both fields. GetIndexOfFirstTxProcessed does not matter when we try to determine if a block is partially executed. Only the index of the last tx executed will say if a miniblock is not yet fully executed.

Reply:
A mini block referenced in a shard header with first index executed set to 0 or 5 could be partially executed but also fully executed, depending of the value of the last index executed.

Reply:
there is a difference between reserved field nil or empty. Should we change the condition to:
```GO
if err != nil || len(miniBlockHeaderReserved) == 0 {
```
The marhaller will output an empty string for a struct with default values as can be seen in this test:
```GO
func Test2(t *testing.T){
    m := &marshal.GogoProtoMarshalizer{}

    b := &block.MiniBlockReserved{}

    buff, _ := m.Marshal(b)

    fmt.Println(buff)
}
```

Reply:
could be changed to len() to accommodate both the nil and empty

Reply:
Actually miniBlockHeaderReserved could be only nil or NOT empty if you check the method m.getMiniBlockHeaderReserved() . There is no marshal but only unmarshal inside the method, so is not the case you underscored above. Anyways if we want this change, a new PR in elrond-go-core should be done. And also this change should be tested very well, as it will affect many parts of the code where the method is called.

```
// GetMiniBlockHeaderReserved returns the MiniBlockHeader Reserved field as a MiniBlockHeaderReserved
func (m *MiniBlockHeader) getMiniBlockHeaderReserved() (*MiniBlockHeaderReserved, error) {
	if len(m.Reserved) > 0 {
		mbhr := &MiniBlockHeaderReserved{}
		err := mbhr.Unmarshal(m.Reserved)
		if err != nil {
			return nil, err
		}

		return mbhr, nil
	}
	return nil, nil
}
```

Reply:
return nil, nil. That is another red flag

Reply:
This is for backwards compatibility as reserved was not used before scheduled miniblocks and here we can not pass the epoch notifier

Reply:
still not agree. It can easily raise panics as we got used to use the code